File: clients/python/src/dynamodb_streams_consumer/record.py
"""Typed DynamoDB Streams change record, decoded from the sidecar wire format.

The sidecar serializes the Rust ``AttrValue`` enum externally-tagged, e.g.
``{"S": "k1"}``, ``{"N": "42"}``, ``{"Bool": true}``, ``"Null"``,
``{"M": {...}}``, ``{"L": [...]}``, ``{"Ss": [...]}``.

Records are exposed in one of two shapes, chosen once at the ``Worker`` level via
``record_format`` (see :class:`~dynamodb_streams_consumer.worker.Worker`):

* ``"native"`` (default) — native Python values (:func:`decode_attr`); numbers
  stay as strings exactly as DynamoDB represents them (lossless, no float
  rounding). Removes the DynamoDB-JSON unmarshalling burden.
* ``"ddb_json"`` — canonical DynamoDB JSON (:func:`to_ddb_json`), the
  ``{"S": ...}`` / ``{"N": ...}`` / ``{"BOOL": ...}`` / ``{"NULL": true}`` /
  ``{"B": <base64>}`` / ``{"SS"|"NS"|"BS": ...}`` shape that ``boto3``'s
  ``TypeDeserializer`` and the AWS SDKs consume. For SDK interop / KCL parity.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

NATIVE = "native"
DDB_JSON = "ddb_json"


def decode_attr(av: Any) -> Any:
    """Decode one wire attribute value into a native Python value."""
    if av == "Null":
        return None
    if not isinstance(av, dict) or len(av) != 1:
        # Be permissive: unknown shapes pass through unchanged.
        return av
    tag, val = next(iter(av.items()))
    if tag == "S":
        return val
    if tag == "N":
        return val  # keep DynamoDB's canonical string form (lossless)
    if tag == "Bool":
        return val
    if tag == "B":
        # Rust serializes bytes as a JSON array of u8.
        return bytes(val)
    if tag == "M":
        return {k: decode_attr(v) for k, v in val.items()}
    if tag == "L":
        return [decode_attr(v) for v in val]
    if tag == "Ss":
        return list(val)
    if tag == "Ns":
        return list(val)  # numeric set members stay strings
    if tag == "Bs":
        return [bytes(b) for b in val]
    return av


def decode_item(item: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if not item:
        return {}
    return {k: decode_attr(v) for k, v in item.items()}


def to_ddb_json(av: Any) -> Any:
    """Convert one wire attribute value into canonical DynamoDB JSON.

    Maps the sidecar's externally-tagged form to the standard DynamoDB attribute
    value shape (the form ``boto3.dynamodb.types.TypeDeserializer`` and the
    low-level SDK accept), so a processor can hand it straight to the SDK."""
    if av == "Null":
        return {"NULL": True}
    if not isinstance(av, dict) or len(av) != 1:
        return av
    tag, val = next(iter(av.items()))
    if tag == "S":
        return {"S": val}
    if tag == "N":
        return {"N": val}
    if tag == "Bool":
        return {"BOOL": val}
    if tag == "B":
        return {"B": base64.b64encode(bytes(val)).decode("ascii")}
    if tag == "M":
        return {"M": {k: to_ddb_json(v) for k, v in val.items()}}
    if tag == "L":
        return {"L": [to_ddb_json(v) for v in val]}
    if tag == "Ss":
        return {"SS": list(val)}
    if tag == "Ns":
        return {"NS": list(val)}
    if tag == "Bs":
        return {"BS": [base64.b64encode(bytes(b)).decode("ascii") for b in val]}
    return av


def ddb_json_item(item: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if not item:
        return {}
    return {k: to_ddb_json(v) for k, v in item.items()}


@dataclass
class Record:
    """One item-level change delivered from a DynamoDB stream shard."""

    shard_id: str
    sequence_number: Optional[str]
    event_name: Optional[str]  # INSERT / MODIFY / REMOVE
    stream_view_type: Optional[str]
    keys: Dict[str, Any] = field(default_factory=dict)
    new_image: Optional[Dict[str, Any]] = None
    old_image: Optional[Dict[str, Any]] = None

    @classmethod
    def from_wire(
        cls, shard_id: str, wire: Dict[str, Any], record_format: str = NATIVE
    ) -> "Record":
        ni = wire.get("new_image")
        oi = wire.get("old_image")
        conv_item = ddb_json_item if record_format == DDB_JSON else decode_item
        return cls(
            shard_id=shard_id,
            sequence_number=wire.get("sequence_number"),
            event_name=wire.get("event_name"),
            stream_view_type=wire.get("stream_view_type"),
            keys=conv_item(wire.get("keys")),
            new_image=conv_item(ni) if ni else None,
            old_image=conv_item(oi) if oi else None,
        )

File: clients/java/src/main/java/com/amazon/dynamodbstreams/consumer/Record.java
package com.amazon.dynamodbstreams.consumer;

import java.util.Collections;
import java.util.Map;

/**
 * A DynamoDB Streams change record delivered to a {@link RecordProcessor}.
 *
 * <p>Item images ({@link #keys()}, {@link #newImage()}, {@link #oldImage()}) are
 * decoded according to the worker's {@link RecordFormat}: in
 * {@link RecordFormat#NATIVE} each value is a plain Java object
 * ({@link String}, {@link Boolean}, {@code null}, {@code byte[]},
 * {@link java.util.List}, or nested {@code Map<String, Object>}); in
 * {@link RecordFormat#DDB_JSON} each value is a single-entry map in canonical
 * DynamoDB JSON form (e.g. {@code {"S": "x"}}).
 */
public final class Record {
    private final String shardId;
    private final String eventName;
    private final String sequenceNumber;
    private final String streamViewType;
    private final Map<String, Object> keys;
    private final Map<String, Object> newImage;
    private final Map<String, Object> oldImage;

    public Record(String shardId, String eventName, String sequenceNumber, String streamViewType,
                  Map<String, Object> keys, Map<String, Object> newImage, Map<String, Object> oldImage) {
        this.shardId = shardId;
        this.eventName = eventName;
        this.sequenceNumber = sequenceNumber;
        this.streamViewType = streamViewType;
        this.keys = keys == null ? Collections.emptyMap() : keys;
        this.newImage = newImage;
        this.oldImage = oldImage;
    }

    /** The shard this record was delivered from. */
    public String shardId() {
        return shardId;
    }

    /** INSERT / MODIFY / REMOVE. */
    public String eventName() {
        return eventName;
    }

    /** The record's sequence number. */
    public String sequenceNumber() {
        return sequenceNumber;
    }

    /** KEYS_ONLY / NEW_IMAGE / OLD_IMAGE / NEW_AND_OLD_IMAGES. */
    public String streamViewType() {
        return streamViewType;
    }

    /** The key attributes of the changed item. */
    public Map<String, Object> keys() {
        return keys;
    }

    /** The item image after the change, or {@code null} when absent. */
    public Map<String, Object> newImage() {
        return newImage;
    }

    /** The item image before the change, or {@code null} when absent. */
    public Map<String, Object> oldImage() {
        return oldImage;
    }
}
